numpy/random/mtrand/mtrand.pyx

```diff
@@ -4652,33 +4652,33 @@ cdef class RandomState:
                [0, 1, 2]])
 
         """
-        cdef npy_intp i, j
-
-        i = len(x) - 1
+        cdef:
+            npy_intp[::1] idxs
+            npy_intp i, j, n = len(x)
 
         # Logic adapted from random.shuffle()
-        if isinstance(x, np.ndarray) and \
-           (x.ndim > 1 or x.dtype.fields is not None):
-            # For a multi-dimensional ndarray, indexing returns a view onto
-            # each row. So we can't just use ordinary assignment to swap the
-            # rows; we need a bounce buffer.
-            buf = np.empty_like(x[0])
-            with self.lock:
-                while i > 0:
-                    j = rk_interval(i, self.internal_state)
-                    buf[...] = x[j]
-                    x[j] = x[i]
-                    x[i] = buf
-                    i = i - 1
+        if isinstance(x, np.ndarray):
+            if x.ndim == 1 and x.dtype.itemsize == np.dtype(np.intp).itemsize:
+                # Directly shuffle the array if possible.
+                self._shuffle_intpsized(x.view(np.intp))
+            else:
+                # Take from a shuffled range to benefit from static typing.
+                idxs = np.arange(n, dtype=np.intp)
+                self._shuffle_intpsized(idxs)
+                x.take(idxs, 0, out=x)
```

Comment:
I doubt `np.take` does not work in-place or checks this. Which means the result can be corrupted by memory layout. I think you have to create a copy unfortunately.
Depends a bit on the shape, but unless your array is `(N, m)` with `m` being quite small, indexing is probably faster, but does not matter much.